### PR TITLE
feat(matches): Implement Service Issue-Repository Matches

### PR DESCRIPTION
**Pull request by dorneanu**
## Description
This PR implements #188. 

1) The `EventRegistry` has been changed to use an `EventHandler` (interface) and `EventHandlerFunc` (as in [net/http](https://pkg.go.dev/net/http#Handler)). This way we can still `On<event name>` functions as handlers whenever an event occurs. 

2) We added extra logic whenever a new entity is created

- a new `Service`
  - for each IssueRepository add a default priority
- a new `IssueRepository`
  - for each Service the default priority will be set

## What type of PR is this? (check all applicable)

- [x] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [x] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [x] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] 📦 Chore (Release)
- [ ] ⏩ Revert

## Related Tickets & Documents

- Related Issue #188 

## Added tests?

- [x] 👍 yes
- [ ] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help
- [ ] Separate ticket for tests # (issue/pr)

Added tests in 

- service_handler_test.go
- issue_repository_test.go

## Added to documentation?

- [ ] 📜 README.md
- [ ] 🤝 Documentation pages updated
- [x] 🙅 no documentation needed
- [ ] (if applicable) generated OpenAPI docs for CRD changes

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: internal/app/issue_match/issue_match_handler_events.go [hunk 1]
```diff
@@ -78,3 +83,204 @@ type RemoveEvidenceFromIssueMatchEvent struct {
 func (e *RemoveEvidenceFromIssueMatchEvent) Name() event.EventName {
 	return RemoveEvidenceFromIssueMatchEventName
 }
+
+func OnComponentInstanceCreate(db database.Database, event event.Event) {
```
Comment (drochow) [inline internal/app/issue_match/issue_match_handler_events.go:87]: Isnt that the next feature? Issue Mapping/Matching?

File: internal/app/heureka.go [hunk 2]
```diff
@@ -69,6 +69,26 @@ func NewHeurekaApp(db database.Database) *HeurekaApp {
 	}
 }
 
+func (h *HeurekaApp) SubscribeHandlers() {
+	// Event handlers for ComponentInstance
+	h.eventRegistry.RegisterEventHandler(
```
Comment (drochow) [inline internal/app/heureka.go:74]: Isnt that the Issue Matching feature?

**Review by drochow — DISMISSED** [review]
Good Job, not sure if the issue matching stuff is added on accident?

File: internal/app/issue_match/issue_match_handler_events.go (hunk 1 above)
Comment (dorneanu) [reply internal/app/issue_match/issue_match_handler_events.go:87]: I've "spined off" this branch from `basic_matching`. I think this function was already there before.

File: internal/app/heureka.go (hunk 2 above)
Comment (dorneanu) [reply internal/app/heureka.go:74]: Same as [here](https://github.com/cloudoperators/heureka/pull/209#discussion_r1762536414)

**Review by drochow — APPROVED** [review]

**Review by drochow — APPROVED** [review]

File: internal/app/issue_repository/issue_repository_handler_events.go [hunk 3]
```diff
@@ -48,3 +50,50 @@ type DeleteIssueRepositoryEvent struct {
 func (e *DeleteIssueRepositoryEvent) Name() event.EventName {
 	return DeleteIssueRepositoryEventName
 }
+
+// OnIssueRepositoryCreate is a handler for the CreateIssueRepositoryEvent
+// Is adding the default priority  for the default issue repository
+func OnIssueRepositoryCreate(db database.Database, e event.Event) {
+	// TODO: make this configureable
+	var defaultPrio int64 = 100
```
Comment (MR2011) [inline internal/app/issue_repository/issue_repository_handler_events.go:58]: Should we fix that todo now?

File: internal/app/issue_repository/issue_repository_handler_events.go [hunk 4]
```diff
@@ -48,3 +50,50 @@ type DeleteIssueRepositoryEvent struct {
...
+	var defaultRepoName string = "nvd"
```
Comment (MR2011) [inline internal/app/issue_repository/issue_repository_handler_events.go:59]: This is not required for IssueRepositoryCreate, right?

File: internal/app/service/service_handler_events.go [hunk 5]
```diff
@@ -109,3 +111,47 @@ type RemoveIssueRepositoryFromServiceEvent struct {
 func (e *RemoveIssueRepositoryFromServiceEvent) Name() event.EventName {
 	return RemoveIssueRepositoryFromServiceEventName
 }
+
+// OnServiceCreate is a handler for the CreateServiceEvent
+// Is creating a single default priority for the default issue repository
+func OnServiceCreate(db database.Database, e event.Event) {
+	//@todo configure somewhere
+	var defaultPrio int64 = 100
```
Comment (MR2011) [inline internal/app/service/service_handler_events.go:119]: Same here: Do we want to fix that todo now?

File: internal/app/issue_repository/issue_repository_handler_events.go (hunk 4 above)
Comment (dorneanu) [reply internal/app/issue_repository/issue_repository_handler_events.go]: You're right. I've removed it.

File: internal/app/service/service_handler_events.go (hunk 5 above)
Comment (dorneanu) [reply internal/app/service/service_handler_events.go]: Implemented.

File: internal/app/issue_repository/issue_repository_handler_events.go (hunk 3 above)
Comment (dorneanu) [reply internal/app/issue_repository/issue_repository_handler_events.go]: Implemented.

File: internal/database/mariadb/database.go [hunk 6]
```diff
@@ -73,7 +74,8 @@ func NewSqlDatabase(cfg util.Config) (*SqlDatabase, error) {
 	}
 	db.Exec(fmt.Sprintf("USE %s", cfg.DBName))
 	return &SqlDatabase{
-		db: db,
+		db:     db,
+		config: &cfg,
```
Comment (MR2011) [inline internal/database/mariadb/database.go:78]: Do we want to pass the whole config to the database? It does not only contain database configuration WDY @drochow @dorneanu

**Review by drochow — APPROVED** [review]
LGTM

**Comment by MR2011** [comment]
The default repository name should also be defined in an environment variable

**Review by MR2011 — APPROVED** [review]

**Review by drochow — APPROVED** [review]

**Pull request merged**